### ISSUE TITLE
cppcoreguidelines-avoid-non-const-global-variables is stricter than the guidelines

### ISSUE THREAD

**Issue by pgerell** [llvm/llvm-project#47384]
|  |  |
| --- | --- |
| Bugzilla Link | [48040](https://llvm.org/bz48040) |
| Version | unspecified |
| OS | All |

## Extended Description 
The text of the guidelines proposes the following enforcement:

(Simple) Report all non-const variables declared at namespace scope and global pointers/references to non-const data.

clang-tidy however complains on this code snippet even though S::i is not declared at namespace scope.

struct S {
private:
    static int i;
};
int S::i = 0;

https://godbolt.org/z/G7bW5d

**Comment by davidhunter22** [comment]
I think the above example makes cppcoreguidelines-avoid-non-const-global-variables pretty much unusable. It basically implies that private mutable static members are not allowed which to me seems wierd, what are you meant to do instead?

**Comment by Dooez** [comment]
Also happens when variable is in an unnamed namespace;
namespace named {
namespace {
int i;
}
} 

https://godbolt.org/z/hxGYrM7Kc

**Comment by gnbond** [comment]
The same (not surprisingly) occurs with static members of class templates:

```
template<typename T>
struct S {
private:
    static T i;
};
template<typename T>
T S<T>::i{};
```
https://godbolt.org/z/zWbbWf3Y1

For us, this causes so many false positives that the rule is effectively useless and we have turned it off.  And it (or at least the description) is also _wrong_: These are NOT "globally accessible" variables.

**Comment by ZiperRom1** [comment]
Is it done on purpose to avoid some kind on Singleton pattern ? (I know many people find it evil XD)

```c++
namespace Logger {
    class Quill {
        public:
            Quill() = delete;

            static auto getLogger() -> quill::Logger*;

        private:
            static quill::Logger* _instance;  // <--- This line warns

            static auto _initLogger() -> void;
    };
}  // namespace Logger
```

**Comment by bjosv** [comment]
> Is it done on purpose to avoid some kind on Singleton pattern ? (I know many people find it evil XD)

I have seen a mention in https://github.com/isocpp/CppCoreGuidelines/issues/1599#issuecomment-656277299 that singletons would be covered by the rule I.3.
This checker only mentions I.2 and R.6 in the [docs](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines/avoid-non-const-global-variables.html), so it's not crystal clear..

**Comment by PiotrZSL** [comment]
https://reviews.llvm.org/D157180
